File: src-tauri/src/device.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeviceInfo {
    pub serial: String,
    pub mode: DeviceMode,
    pub properties: Option<DeviceProperties>,
    pub connected: bool,
    pub last_seen: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum DeviceMode {
    Sys,          // 系统模式
    Rec,          // Recovery模式
    Fastboot,     // Fastboot模式
    Fastbootd,    // Fastbootd模式
    Sideload,     // Sideload模式
    Edl,          // EDL模式
    Unauthorized, // 未授权
    Offline,      // 离线
    Unknown,      // 未知
}

impl DeviceMode {
    pub fn from_adb_status(status: &str) -> Self {
        match status.trim() {
            "device" => DeviceMode::Sys,
            "recovery" => DeviceMode::Rec,
            "sideload" => DeviceMode::Sideload,
            "unauthorized" => DeviceMode::Unauthorized,
            "offline" => DeviceMode::Offline,
            _ => DeviceMode::Unknown,
        }
    }

    #[allow(dead_code)]
    pub fn from_fastboot_status(_status: &str) -> Self {
        // Fastboot通常只返回设备序列号，需要进一步检测是fastboot还是fastbootd
        DeviceMode::Fastboot
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct DeviceProperties {
    // 设备基本信息
    pub market_name: Option<String>,   // ro.product.marketname
    pub product_name: Option<String>,  // ro.product.name
    pub brand: Option<String>,         // ro.product.brand
    pub model: Option<String>,         // ro.product.model
    pub device_name: Option<String>,   // ro.product.device
    pub manufacturer: Option<String>,  // ro.product.manufacturer
    pub serial_number: Option<String>, // ro.serialno

    // 系统版本信息
    pub android_version: Option<String>, // ro.build.version.release
    pub sdk_version: Option<String>,     // ro.build.version.sdk
    pub build_id: Option<String>,        // ro.build.id
    pub build_display_id: Option<String>, // ro.build.display.id
    pub system_version: Option<String>,  // ro.system.build.version.incremental
    pub security_patch_level: Option<String>, // ro.build.version.security_patch
    pub build_fingerprint: Option<String>, // ro.build.fingerprint
    pub build_date: Option<String>,      // ro.build.date
    pub build_user: Option<String>,      // ro.build.user
    pub build_host: Option<String>,      // ro.build.host
    pub miui_version: Option<String>,    // ro.mi.ui.version.name



    // 硬件信息
    pub cpu_abi: Option<String>,           // ro.product.cpu.abi
    pub cpu_abi_list: Option<String>,      // ro.product.cpu.abilist
    pub soc_manufacturer: Option<String>,  // ro.soc.manufacturer
    pub soc_model: Option<String>,         // ro.soc.model
    pub hardware: Option<String>,          // ro.hardware
    pub hardware_chipname: Option<String>, // ro.hardware.chipname
    pub board_platform: Option<String>,    // ro.board.platform
    pub product_board: Option<String>,     // ro.product.board

    // 安全和启动信息
    pub bootloader_locked: Option<bool>, // ro.boot.flash.locked
    pub verified_boot_state: Option<String>, // ro.boot.verifiedbootstate
    pub verity_mode: Option<String>,     // ro.boot.veritymode
    pub debuggable: Option<bool>,        // ro.debuggable
    pub secure: Option<bool>,            // ro.secure
    pub adb_secure: Option<bool>,        // ro.adb.secure

    // 显示和UI信息
    pub lcd_density: Option<String>, // ro.sf.lcd_density
    pub locale: Option<String>,      // ro.product.locale
    pub timezone: Option<String>,    // persist.sys.timezone

    // 网络和通信
    pub default_network: Option<String>, // ro.telephony.default_network
    pub first_api_level: Option<String>, // ro.product.first_api_level
    pub vndk_version: Option<String>,    // ro.vndk.version

    // 运行时信息
    pub imei: Option<String>,              // 需要特殊权限获取
    pub battery_level: Option<i32>,        // dumpsys battery
    pub screen_resolution: Option<String>, // wm size
    pub total_memory: Option<String>,      // /proc/meminfo
    pub available_storage: Option<String>, // df /data

    // 详细内存和存储信息
    pub memory_total: Option<u64>,      // 总内存 (MB)
    pub memory_used: Option<u64>,       // 已使用内存 (MB)
    pub memory_available: Option<u64>,  // 可用内存 (MB)
    pub storage_total: Option<u64>,     // 总存储 (MB)
    pub storage_used: Option<u64>,      // 已使用存储 (MB)
    pub storage_available: Option<u64>, // 可用存储 (MB)
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommandResult {
    pub success: bool,
    pub output: String,
    pub error: Option<String>,
    pub exit_code: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AdbCommand {
    pub command: String,
    pub args: Vec<String>,
    pub timeout: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileTransferProgress {
    pub id: String,
    pub file_name: String,
    pub total_size: u64,
    pub transferred: u64,
    pub progress: f64,
    pub speed: Option<String>,
    pub eta: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InstalledApp {
    pub package_name: String,
    pub app_name: Option<String>,
    pub version_name: Option<String>,
    pub version_code: Option<String>,
    pub install_location: Option<String>,
    pub is_system_app: bool,
    pub is_enabled: bool,
    pub apk_path: Option<String>,
    pub install_time: Option<String>,
    pub update_time: Option<String>,
    pub permissions: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ApkInfo {
    pub file_path: String,
    pub package_name: Option<String>,
    pub app_name: Option<String>,
    pub version_name: Option<String>,
    pub version_code: Option<String>,
    pub min_sdk_version: Option<String>,
    pub target_sdk_version: Option<String>,
    pub compile_sdk_version: Option<String>,
    pub permissions: Vec<String>,
    pub features: Vec<String>,
    pub file_size: u64,
    pub is_debuggable: bool,
    pub is_test_only: bool,
    pub icon_path: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InstallProgress {
    pub id: String,
    pub file_name: String,
    pub package_name: Option<String>,
    pub status: InstallStatus,
    pub progress: f64,
    pub message: Option<String>,
    pub start_time: chrono::DateTime<chrono::Utc>,
    pub end_time: Option<chrono::DateTime<chrono::Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum InstallStatus {
    Pending,
    Installing,
    Success,
    Failed,
    Cancelled,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BatchOperation {
    pub id: String,
    pub operation_type: BatchOperationType,
    pub total_items: usize,
    pub completed_items: usize,
    pub failed_items: usize,
    pub status: BatchOperationStatus,
    pub items: Vec<BatchOperationItem>,
    pub start_time: chrono::DateTime<chrono::Utc>,
    pub end_time: Option<chrono::DateTime<chrono::Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum BatchOperationType {
    Install,
    Uninstall,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum BatchOperationStatus {
    Pending,
    Running,
    Completed,
    Failed,
    Cancelled,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BatchOperationItem {
    pub id: String,
    pub name: String,
    pub status: InstallStatus,
    pub message: Option<String>,
}

impl DeviceInfo {
    pub fn new(serial: String, mode: DeviceMode) -> Self {
        Self {
            serial,
            mode,
            properties: None,
            connected: true,
            last_seen: Some(Utc::now()),
        }
    }

    #[allow(dead_code)]
    pub fn update_properties(&mut self, properties: DeviceProperties) {
        self.properties = Some(properties);
        self.last_seen = Some(Utc::now());
    }

    pub fn is_adb_available(&self) -> bool {
        matches!(
            self.mode,
            DeviceMode::Sys | DeviceMode::Rec | DeviceMode::Sideload
        )
    }

    pub fn is_fastboot_available(&self) -> bool {
        matches!(self.mode, DeviceMode::Fastboot | DeviceMode::Fastbootd)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeviceFile {
    pub name: String,
    pub path: String,
    pub is_directory: bool,
    pub size: Option<u64>,
    pub permissions: Option<String>,
    pub modified_time: Option<String>,
}

File: src/types/device.ts
export interface DeviceInfo {
  serial: string;
  mode: DeviceMode;
  properties?: DeviceProperties;
  connected: boolean;
  lastSeen?: Date;
  boardSerialNumber?: string;
  fastbootVariables?: Record<string, string>; // Fastboot模式下的设备变量
}

export type DeviceMode = 
  | "sys"        // 系统模式
  | "rec"        // Recovery模式
  | "fastboot"   // Fastboot模式
  | "fastbootd"  // Fastbootd模式
  | "sideload"   // Sideload模式
  | "edl"        // EDL模式
  | "unauthorized" // 未授权
  | "offline"    // 离线
  | "unknown";   // 未知

export interface DeviceProperties {
  // 设备基本信息
  marketName?: string;        // 商品名称
  productName?: string;       // 产品名称
  brand?: string;            // 品牌
  model?: string;            // 型号
  deviceName?: string;       // 设备代号
  manufacturer?: string;     // 制造商
  serialNumber?: string;     // 序列号

  // 系统版本信息
  androidVersion?: string;   // Android版本
  sdkVersion?: string;       // SDK版本
  buildId?: string;          // 构建ID
  buildDisplayId?: string;   // 构建显示ID
  systemVersion?: string;    // 系统版本
  securityPatchLevel?: string; // 安全补丁级别
  buildFingerprint?: string;  // 构建指纹
  buildDate?: string;        // 构建日期
  buildUser?: string;        // 构建用户
  buildHost?: string;        // 构建主机
  miuiVersion?: string;      // MIUI版本

  // 硬件信息
  cpuAbi?: string;          // CPU架构
  cpuAbiList?: string;      // 支持的CPU架构列表
  socManufacturer?: string;  // SoC制造商
  socModel?: string;        // SoC型号
  hardware?: string;        // 硬件平台
  hardwareChipname?: string; // 硬件芯片名称
  boardPlatform?: string;   // 主板平台
  productBoard?: string;    // 产品主板

  // 安全和启动信息
  bootloaderLocked?: boolean; // Bootloader锁定状态
  verifiedBootState?: string; // 验证启动状态
  verityMode?: string;      // 完整性验证模式
  debuggable?: boolean;     // 调试模式
  secure?: boolean;         // 安全模式
  adbSecure?: boolean;      // ADB安全模式

  // 显示和UI信息
  lcdDensity?: string;      // 屏幕密度
  locale?: string;          // 语言区域
  timezone?: string;        // 时区

  // 网络和通信
  defaultNetwork?: string;  // 默认网络类型
  firstApiLevel?: string;   // 首次API级别
  vndkVersion?: string;     // VNDK版本

  // 运行时信息
  imei?: string;            // IMEI
  batteryLevel?: number;    // 电池电量
  screenResolution?: string; // 屏幕分辨率
  totalMemory?: string;     // 总内存
  availableStorage?: string; // 可用存储
  parallelDownloadFlash?: boolean; // 并行刷写支持
  offModeCharge?: boolean;  // 关机充电模式
  antiRollback?: boolean;   // 防回滚保护
  cpuid?: string;           // CPU唯一ID
}

export interface DeviceStatus {
  devices: DeviceInfo[];
  selectedDevice?: DeviceInfo;
  isScanning: boolean;
  lastUpdate: Date;
}

export interface AdbCommand {
  command: string;
  args?: string[];
  timeout?: number;
}

export interface CommandResult {
  success: boolean;
  output: string;
  error?: string;
  exitCode?: number;
}

export interface InstalledApp {
  packageName: string;
  versionName?: string;
  versionCode?: string;
  installLocation?: string;
  isSystemApp: boolean;
  isEnabled: boolean;
  apkPath?: string;
  installTime?: string;
  updateTime?: string;
  permissions: string[];
}

export interface ApkInfo {
  filePath: string;
  packageName?: string;
  appName?: string;
  versionName?: string;
  versionCode?: string;
  minSdkVersion?: string;
  targetSdkVersion?: string;
  compileSdkVersion?: string;
  permissions: string[];
  features: string[];
  fileSize: number;
  isDebuggable: boolean;
  isTestOnly: boolean;
  iconPath?: string;
}

export interface InstallProgress {
  id: string;
  fileName: string;
  packageName?: string;
  status: InstallStatus;
  progress: number;
  message?: string;
  startTime: string;
  endTime?: string;
}

export type InstallStatus = 'pending' | 'installing' | 'success' | 'failed' | 'cancelled';

export interface BatchOperation {
  id: string;
  operationType: BatchOperationType;
  totalItems: number;
  completedItems: number;
  failedItems: number;
  status: BatchOperationStatus;
  items: BatchOperationItem[];
  startTime: string;
  endTime?: string;
}

export type BatchOperationType = 'install' | 'uninstall' | 'freeze' | 'unfreeze' | 'forceStop' | 'exportApk' | 'clearData';
export type BatchOperationStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';

export interface BatchOperationItem {
  id: string;
  name: string;
  status: InstallStatus;
  message?: string;
}

export interface ConnectionInfo {
  serial: string;
  state: string;
  connected: boolean;
  usb_connection: boolean;
  wifi_connection: boolean;
  adb_version?: string;
  transport_id?: string;
}

// APK市场相关类型定义
export interface ApkMarketData {
  apk_list: ApkCategory[];
}

export interface ApkCategory {
  category: string;
  apps: ApkItem[];
}

export interface ApkItem {
  name: string;
  download_link: {
    url: string;
    is_direct: boolean;
  };
}

export interface ApkDownloadItem {
  id: string;
  name: string;
  category: string;
  url: string;
  isDirect: boolean;
  status: ApkDownloadStatus;
  progress: number;
  downloadedSize: number;
  totalSize: number;
  filePath?: string;
  error?: string;
  startTime?: string;
  endTime?: string;
}

export type ApkDownloadStatus = 'pending' | 'downloading' | 'completed' | 'failed' | 'paused' | 'cancelled';

export interface DeviceFile {
  name: string;
  path: string;
  isDirectory: boolean;
  size?: number;
  permissions?: string;
  modifiedTime?: string;
}
